drivers/vhost/dsp-sof.c

```diff
@@ -0,0 +1,968 @@
+{
+	struct snd_pcm_substream *substream;
+	struct snd_sof_pcm *spcm;
+	struct snd_soc_pcm_runtime *rtd = NULL;
+	struct sof_ipc_stream *stream;
+	u32 comp_id;
+	int direction;
+
+	stream = (struct sof_ipc_stream *)ipc_data;
+	comp_id = stream->comp_id;
+
+	spcm = dsp_sof_find_spcm_comp(sdev, comp_id, &direction);
+	if (!spcm)
+		return 0;
+
+	substream = dsp_sof_get_substream(sdev, &rtd, direction);
+	if (substream) {
+		snd_sof_pcm_platform_close(sdev, substream);
+
+		/* TODO: codec close */
+
+		substream->ref_count = 0;
+		if (substream->runtime) {
+			free_pages_exact(substream->runtime->status,
+					 PAGE_ALIGN(sizeof(struct snd_pcm_mmap_status)));
+			free_pages_exact(substream->runtime->control,
+					 PAGE_ALIGN(sizeof(struct snd_pcm_mmap_control)));
+			kfree(substream->runtime);
+			substream->runtime = NULL;
+		}
+	}
+
+	list_del(&spcm->list);
+	kfree(spcm);
+
+	return 0;
+}
+
+static int dsp_sof_ipc_stream_capture(struct snd_sof_pcm_stream *stream,
+				      struct snd_pcm_runtime *runtime,
+				      struct dsp_sof_data_req *req,
+				      struct dsp_sof_data_resp *reply)
+{
+	size_t data_size = req->size;
+	int ret;
+
+	stream->guest_offset = req->offset;
+
+	if (req->offset + data_size > runtime->dma_bytes) {
+		reply->size = 0;
+		ret = -ENOBUFS;
+	} else {
+		stream->guest_offset += data_size;
+
+		memcpy(reply->data, runtime->dma_area + req->offset, data_size);
+		reply->size = data_size;
+		ret = 0;
+	}
+
+	reply->error = ret;
+
+	return ret;
+}
+
+static int dsp_sof_ipc_stream_playback(struct snd_sof_pcm_stream *stream,
+				       struct snd_pcm_runtime *runtime,
+				       struct dsp_sof_data_req *req,
+				       struct dsp_sof_data_resp *reply)
+{
+	size_t data_size = req->size;
+	int ret;
+
+	stream->guest_offset = req->offset;
+
+	if (req->offset + data_size > runtime->dma_bytes) {
+		ret = -ENOBUFS;
+	} else {
+		stream->guest_offset += data_size;
+
+		memcpy(runtime->dma_area + req->offset, req->data, data_size);
+		ret = 0;
+	}
+
+	reply->error = ret;
+	reply->size = 0;
+
+	return ret;
+}
+
+int dsp_sof_ipc_stream_data(struct snd_sof_dev *sdev,
+			    struct dsp_sof_data_req *req,
+			    struct dsp_sof_data_resp *reply)
+{
+	struct snd_soc_pcm_runtime *rtd;
+	int direction;
+	struct snd_sof_pcm *spcm = dsp_sof_find_spcm_comp(sdev,
+						req->comp_id, &direction);
+	struct snd_pcm_substream *substream = dsp_sof_get_substream(sdev, &rtd,
+								    direction);
+
+	if (!spcm || !substream) {
+		reply->error = -ENODEV;
+		reply->size = 0;
+		return reply->error;
+	}
+
+	if (direction == SNDRV_PCM_STREAM_PLAYBACK)
+		return dsp_sof_ipc_stream_playback(spcm->stream + direction,
+						substream->runtime, req, reply);
+
+	return dsp_sof_ipc_stream_capture(spcm->stream + direction,
+					  substream->runtime, req, reply);
+}
+
+/* handle the stream ipc */
+static int dsp_sof_ipc_stream(struct snd_sof_dev *sdev,
+			      struct sof_ipc_cmd_hdr *hdr, void *reply_buf)
```

Comment:
what is all this? This looks like a copy/paste and rename for all existing PCM functions?

Reply:
@plbossart this specific function is also close to the original, I received for this work. I couldn't find what it is copied from?